File: truevfs-kernel/src/test/java/de/schlichtherle/truevfs/kernel/se/CacheEntryTest.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se;

import de.schlichtherle.truevfs.kernel.se.CacheEntry.Strategy;
import static de.schlichtherle.truevfs.kernel.se.CacheEntry.Strategy.WRITE_BACK;
import static de.schlichtherle.truevfs.kernel.se.CacheEntry.Strategy.WRITE_THROUGH;
import static net.truevfs.kernel.cio.Entry.Access.READ;
import static net.truevfs.kernel.cio.Entry.Access.WRITE;
import static net.truevfs.kernel.cio.Entry.Size.DATA;
import static net.truevfs.kernel.cio.Entry.UNKNOWN;
import net.truevfs.kernel.cio.*;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Objects;
import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;
import org.junit.Before;
import org.junit.Test;

/**
 * @author Christian Schlichtherle
 */
public final class CacheEntryTest {

    private static final int INITIAL_CAPACITY = 32;
    private static final String MOCK_ENTRY_NAME = "mock";
    private static final String MOCK_ENTRY_DATA_READ = "read";
    private static final String MOCK_ENTRY_DATA_WRITE = "write";

    private ByteArrayIOPool pool;

    @Before
    public void setUp() throws IOException {
        pool = new ByteArrayIOPool(5);
    }

    @Test
    public void testCaching() throws IOException {
        for (final Strategy strategy : new Strategy[] {
            WRITE_THROUGH,
            WRITE_BACK,
        }) {
            final CacheEntry cache = strategy.newCacheEntry(pool);
            ByteArrayIOBuffer front;
            ByteArrayIOBuffer back;

            back = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            back.setData(MOCK_ENTRY_DATA_READ.getBytes());
            cache   .configure(new BrokenInputSocket(back))
                    .configure(new BrokenOutputSocket(back));
            assertThat(pool.size(), is(0));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));

            front = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            assertThat(front.getData(), nullValue());
            try {
                IOSockets.copy(cache.input(), front.output());
                fail();
            } catch (IOException expected) {
            }
            assertThat(pool.size(), is(0));
            assertThat(front.getData(), nullValue());
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));

            cache   .configure(back.input())
                    .configure(back.output());
            assertThat(pool.size(), is(0));
            assertThat(front.getData(), nullValue());
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));

            front = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            assertThat(pool.size(), is(0));
            assertThat(front.getData(), nullValue());
            IOSockets.copy(cache.input(), front.output());
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_READ.length()));

            front = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            front.setData(MOCK_ENTRY_DATA_WRITE.getBytes());
            cache   .configure(new BrokenInputSocket(back))
                    .configure(new BrokenOutputSocket(back));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_READ.length()));

            try {
                IOSockets.copy(front.input(), cache.output());
                if (WRITE_THROUGH != strategy) {
                    assertThat( back.getCount(WRITE), is(0));
                    cache.flush();
                }
                fail();
            } catch (IOException expected) {
            }
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            cache   .configure(back.input())
                    .configure(back.output());
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            IOSockets.copy(front.input(), cache.output());
            if (WRITE_THROUGH != strategy) {
                assertThat( back.getCount(WRITE), is(0));
                cache.flush();
            }
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(1));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            back = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            back.setData(MOCK_ENTRY_DATA_READ.getBytes());
            cache   .configure(new BrokenInputSocket(back))
                    .configure(new BrokenOutputSocket(back));
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            front = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            IOSockets.copy(cache.input(), front.output());
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            cache.release();
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));
            assertThat(pool.size(), is(0));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));

            front = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            try {
                IOSockets.copy(cache.input(), front.output());
                fail();
            } catch (IOException excepted) {
            }
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));
            assertThat(pool.size(), is(0));
            assertThat(front.getData(), nullValue());
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));

            cache   .configure(back.input())
                    .configure(back.output());
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));
            assertThat(pool.size(), is(0));
            assertThat(front.getData(), nullValue());
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(0));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));

            IOSockets.copy(cache.input(), front.output());
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_READ.length()));

            front = new ByteArrayIOBuffer(MOCK_ENTRY_NAME, INITIAL_CAPACITY);
            front.setData(MOCK_ENTRY_DATA_WRITE.getBytes());
            cache   .configure(new BrokenInputSocket(back))
                    .configure(new BrokenOutputSocket(back));
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_READ.length()));

            try {
                IOSockets.copy(front.input(), cache.output());
                if (WRITE_THROUGH != strategy) {
                    assertThat( back.getCount(WRITE), is(0));
                    cache.flush();
                }
                fail();
            } catch (IOException expected) {
            }
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            cache   .configure(back.input())
                    .configure(back.output());
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_READ));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(0));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            IOSockets.copy(front.input(), cache.output());
            if (WRITE_THROUGH != strategy) {
                assertThat( back.getCount(WRITE), is(0));
                cache.flush();
            }
            assertThat(cache.getSize(DATA), is(not((long) UNKNOWN)));
            assertThat(pool.size(), is(1));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(1));
            assertThat(cache.getSize(DATA), is((long) MOCK_ENTRY_DATA_WRITE.length()));

            cache   .configure(new BrokenInputSocket(back))
                    .configure(new BrokenOutputSocket(back))
                    .release();
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));
            assertThat(pool.size(), is(0));
            assertThat(new String(front.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(new String(back.getData()), equalTo(MOCK_ENTRY_DATA_WRITE));
            assertThat(back.getCount(READ), is(1));
            assertThat(back.getCount(WRITE), is(1));
            assertThat(cache.getSize(DATA), is((long) UNKNOWN));
        }
    }

    private static class BrokenInputSocket
    extends AbstractInputSocket<Entry> {
        private final Entry entry;

        BrokenInputSocket(Entry entry) {
            this.entry = Objects.requireNonNull(entry);
        }

        @Override
        public Entry localTarget() throws IOException {
            return entry;
        }

        @Override
        public InputStream stream() throws IOException {
            return new BrokenInputStream();
        }

        static class BrokenInputStream extends InputStream {

            @Override
            public int read() throws IOException {
                throw new IOException();
            }
        } // class BrokenInputStream
    } // class BrokenInputSocket

    private static class BrokenOutputSocket
    extends AbstractOutputSocket<Entry> {
        private final Entry entry;

        BrokenOutputSocket(Entry entry) {
            this.entry = Objects.requireNonNull(entry);
        }

        @Override
        public Entry localTarget() throws IOException {
            return entry;
        }

        @Override
        public OutputStream stream() throws IOException {
            return new BrokenOutputStream();
        }

        static class BrokenOutputStream extends OutputStream {
            @Override
            public void write(int b) throws IOException {
                throw new IOException();
            }
        } // class BrokenOutputStream
    } // class BrokenOutputSocket
}

File: truevfs-kernel/src/test/scala/de/schlichtherle/truevfs/kernel/se/CacheEntryTest.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se

import de.schlichtherle.truevfs.kernel.se.CacheEntry.Strategy._
import java.io._
import net.truevfs.kernel.cio.Entry.Access._
import net.truevfs.kernel.cio.Entry.Size._
import net.truevfs.kernel.cio.Entry._
import net.truevfs.kernel.cio._
import org.hamcrest.CoreMatchers._
import org.junit.Assert._
import org.junit._
import org.scalatest.junit._

/**
 * @author Christian Schlichtherle
 */
class CacheEntryTest extends AssertionsForJUnit with ShouldMatchersForJUnit {
  import CacheEntryTest._

  private final val initialCapacity = 32
  private final val mockEntryName = "mock"
  private final val mockEntryDataRead = "read"
  private final val mockEntryDataWrite = "write"
  private val pool = new ByteArrayIoPool(5)

  @Test
  def testCaching() {
    for (strategy <- Array[CacheEntry.Strategy](WRITE_THROUGH, WRITE_BACK)) {
      val cache: CacheEntry = strategy.newCacheEntry(pool)
      var front: ByteArrayIoBuffer = null
      var back: ByteArrayIoBuffer = null
      back = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      back.setData(mockEntryDataRead.getBytes)
      cache .configure(new BrokenInputSocket(back))
            .configure(new BrokenOutputSocket(back))
      pool should have size (0)
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(UNKNOWN.asInstanceOf[Long]))
      front = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      front.getData should be (null)
      intercept[IOException] {
        IoSockets.copy(cache.input, front.output)
      }
      pool should have size (0)
      front.getData should be (null)
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(UNKNOWN.asInstanceOf[Long]))
      cache.configure(back.input).configure(back.output)
      pool should have size (0)
      front.getData should be (null)
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(UNKNOWN.asInstanceOf[Long]))
      front = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      pool should have size (0)
      front.getData should be (null)
      IoSockets.copy(cache.input, front.output)
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataRead))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataRead.length.asInstanceOf[Long]))
      front = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      front.setData(mockEntryDataWrite.getBytes)
      cache.configure(new CacheEntryTest.BrokenInputSocket(back)).configure(new CacheEntryTest.BrokenOutputSocket(back))
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataRead.length.asInstanceOf[Long]))
      intercept[IOException] {
        IoSockets.copy(front.input, cache.output)
        if (WRITE_THROUGH ne strategy) {
          assertThat(back.getCount(WRITE), is(0))
          cache.flush()
        }
      }
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      cache.configure(back.input).configure(back.output)
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      IoSockets.copy(front.input, cache.output)
      if (WRITE_THROUGH ne strategy) {
        assertThat(back.getCount(WRITE), is(0))
        cache.flush()
      }
      cache.getSize(DATA) should not be UNKNOWN
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataWrite))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(1))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      back = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      back.setData(mockEntryDataRead.getBytes)
      cache.configure(new CacheEntryTest.BrokenInputSocket(back)).configure(new CacheEntryTest.BrokenOutputSocket(back))
      cache.getSize(DATA) should not be UNKNOWN
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      front = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      IoSockets.copy(cache.input, front.output)
      cache.getSize(DATA) should not be UNKNOWN
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      cache.release()
      cache.getSize(DATA) should be (UNKNOWN)
      pool should have size (0)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      cache.getSize(DATA) should be (UNKNOWN)
      front = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      intercept[IOException] {
        IoSockets.copy(cache.input, front.output)
      }
      cache.getSize(DATA) should be (UNKNOWN)
      pool should have size (0)
      front.getData should be (null)
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      cache.getSize(DATA) should be (UNKNOWN)
      cache.configure(back.input).configure(back.output)
      cache.getSize(DATA) should be (UNKNOWN)
      pool should have size (0)
      front.getData should be (null)
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(0))
      assertThat(back.getCount(WRITE), is(0))
      cache.getSize(DATA) should be (UNKNOWN)
      IoSockets.copy(cache.input, front.output)
      cache.getSize(DATA) should not be (UNKNOWN)
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataRead))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataRead.length.asInstanceOf[Long]))
      front = new ByteArrayIoBuffer(mockEntryName, initialCapacity)
      front.setData(mockEntryDataWrite.getBytes)
      cache.configure(new CacheEntryTest.BrokenInputSocket(back)).configure(new CacheEntryTest.BrokenOutputSocket(back))
      cache.getSize(DATA) should not be (UNKNOWN)
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataRead.length.asInstanceOf[Long]))
      intercept[IOException] {
        IoSockets.copy(front.input, cache.output)
        if (WRITE_THROUGH ne strategy) {
          assertThat(back.getCount(WRITE), is(0))
          cache.flush()
        }
      }
      cache.getSize(DATA) should not be (UNKNOWN)
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      cache.configure(back.input).configure(back.output)
      cache.getSize(DATA) should not be (UNKNOWN)
      assertThat(pool.size, is(1))
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataRead))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(0))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      IoSockets.copy(front.input, cache.output)
      if (WRITE_THROUGH ne strategy) {
        assertThat(back.getCount(WRITE), is(0))
        cache.flush()
      }
      cache.getSize(DATA) should not be (UNKNOWN)
      pool should have size (1)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataWrite))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(1))
      assertThat(cache.getSize(DATA), is(mockEntryDataWrite.length.asInstanceOf[Long]))
      cache .configure(new BrokenInputSocket(back))
            .configure(new BrokenOutputSocket(back))
            .release()
      cache.getSize(DATA) should be (UNKNOWN)
      pool should have size (0)
      assertThat(new String(front.getData), equalTo(mockEntryDataWrite))
      assertThat(new String(back.getData), equalTo(mockEntryDataWrite))
      assertThat(back.getCount(READ), is(1))
      assertThat(back.getCount(WRITE), is(1))
      cache.getSize(DATA) should be (UNKNOWN)
    }
  }
}

private object CacheEntryTest {
  class BrokenInputSocket(override val localTarget: Entry)
  extends AbstractInputSocket[Entry] {
    require(null ne localTarget)

    override def stream = new InputStream {
      override def read = throw new IOException
    }
  }

  class BrokenOutputSocket(override val localTarget: Entry)
  extends AbstractOutputSocket[Entry] {
    require(null ne localTarget)

    override def stream = new OutputStream {
      override def write(b: Int) { throw new IOException }
    }
  }
}
